### PR TITLE
ARROW-11460: [R] Use system libraries if present on Linux

### PR DESCRIPTION
**Pull request by thisisnic**

### CODE REVIEW

**Comment by thisisnic** [comment]
@github-actions crossbow submit test-r-depsource

**Comment by thisisnic** [comment]
@github-actions crossbow submit test-r-depsource

**Comment by thisisnic** [comment]
@github-actions crossbow submit test-r-depsource

**Comment by thisisnic** [comment]
@github-actions crossbow submit test-r-depsource

**Comment by thisisnic** [comment]
@github-actions crossbow submit test-r-depsource

**Comment by thisisnic** [comment]
@github-actions crossbow submit test-r-depsource

File: r/inst/build_arrow_static.sh [hunk 1]
```diff
@@ -45,6 +45,22 @@ else
   ARROW_DEFAULT_PARAM="OFF"
 fi
 
+if [ "$DEPENDENCY_SOURCE" = "AUTO"]; then
```
Comment (jonkeane) [inline r/inst/build_arrow_static.sh:48]:
```suggestion
if [ "$DEPENDENCY_SOURCE" = "AUTO" ]; then
```

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-depsource

**Comment by thisisnic** [comment]
@github-actions crossbow submit test-r-depsource

**Comment by thisisnic** [comment]
@github-actions crossbow submit test-r-depsource

File: r/inst/build_arrow_static.sh [hunk 2]
```diff
@@ -45,6 +45,19 @@ else
...
 
+if [ "$R_STATIC_DEPENDENCY_SOURCE" = "AUTO" ]; then
+  pkg-config --version >/dev/null 2>&1
+  if [ $? -eq 0 ]; then
+    ARROW_DEPENDENCY_SOURCE="AUTO"
+  else
+    ARROW_DEPENDENCY_SOURCE="BUNDLED"
+    echo "**** Warning: $R_STATIC_DEPENDENCY_SOURCE set to 'AUTO' but pkg-config not detected"
+    echo "**** Using bundled dependencies instead"
+  fi
```
Comment (jonkeane) [inline r/inst/build_arrow_static.sh:56]: I don't think it makes sense to try and concoct a CI configuration that hits this else, but I'm curious if you've tried to do it locally and check that you get the expected behavior?

Comment (thisisnic) [reply r/inst/build_arrow_static.sh:56]: I have not but happy to do so to test it. What do you think is the most straightforward approach?

Comment (thisisnic) [reply r/inst/build_arrow_static.sh:56]: Should I just spin up a docker image?

Comment (jonkeane) [reply r/inst/build_arrow_static.sh:56]: For this I would `ARROW_R_DEV=TRUE R_STATIC_DEPENDENCY_SOURCE=AUTO docker-compose build r` and then interactively enter the docker container that is built there, uninstall pkg-config and then run `/arrow/ci/scripts/r_test.sh` or try installing manually

Comment (jonkeane) [reply r/inst/build_arrow_static.sh:56]: Oops, didn't see your second comment before I sent mine, yeah, exactly

File: .env [hunk 3]
```diff
@@ -67,6 +67,7 @@ R_ORG=rhub
 R_IMAGE=ubuntu-gcc-release
 R_TAG=latest
 TZ=UTC
+R_STATIC_DEPENDENCY_SOURCE=BUNDLED
```
Comment (nealrichardson) [inline .env:70]: I don't think we need this here. We have to put env vars here that affect what the Dockerfiles build, but this is a runtime env var, so we can just pass it with `-e` and don't need to declare it here.

File: r/inst/build_arrow_static.sh [hunk 4]
```diff
@@ -54,7 +67,7 @@ ${CMAKE} -DARROW_BOOST_USE_SHARED=OFF \
     -DARROW_COMPUTE=ON \
     -DARROW_CSV=ON \
     -DARROW_DATASET=${ARROW_DATASET:-ON} \
-    -DARROW_DEPENDENCY_SOURCE=BUNDLED \
+    -DARROW_DEPENDENCY_SOURCE=${ARROW_DEPENDENCY_SOURCE} \
```
Comment (jonkeane) [inline r/inst/build_arrow_static.sh:70]: Hmmm, something is going on and this isn't getting passed through, it looks like `DEPENDENCY_SOURCE` is still set to bundled:

https://github.com/ursacomputing/crossbow/runs/3170999772?check_suite_focus=true#step:8:283

I'm still tracking down where this is falling out though, so don't yet have an explanation

File: r/inst/build_arrow_static.sh [hunk 5]
```diff
@@ -45,6 +45,19 @@ else
   ARROW_DEFAULT_PARAM="OFF"
 fi
 
+if [ "$R_STATIC_DEPENDENCY_SOURCE" = "AUTO" ]; then
```
Comment (nealrichardson) [inline r/inst/build_arrow_static.sh:48]: This isn't quite right/complete, and I think if we look at what is required to be complete, we'll see that we don't actually want this logic in this script. Working backwards, the point of Kou's change is that the pkgconfig files that the arrow C++ build does now contain all of the `-l` libs that the build requires from the system (which shouldn't be any if we're doing a BUNDLED build, but with AUTO there will be libs in it). So we need to make sure the `PKG_LIBS` env var in `r/configure` gets them. And since this script, and even `r/tools/nixlibs.R`, can't return anything or set anything in the env that `r/configure` is running in, we need to call `pkg-config` on the arrow C++ install dir from `r/configure`. (This is the missing par.) But that will only work if `pkg-config` exists, so we need to check for it like you do here--but I think we need to check for it in `r/configure` so that that script know whether to set AUTO or BUNDLED.

**Review by nealrichardson — CHANGES_REQUESTED** [review]

File: .env (hunk 3 above)
Comment (jonkeane) [reply .env:70]: To add to that, here's one example of how that's done for a different CI job: https://github.com/apache/arrow/blob/master/.github/workflows/r-without-arrow.yml#L76

File: r/inst/build_arrow_static.sh (hunk 4 above)
Comment (thisisnic) [reply r/inst/build_arrow_static.sh:70]: That's weird; I did a diff comparing the `BUNDLED` and `AUTO` jobs, and the main thing that stood out to me there was that in the `AUTO` one, there's a load of lines that start with "Looking for" and then whatever dependency, so I assumed it was working.  

Could this be related to @nealrichardson 's comment below about it needing to be done earlier, in `r/configure`?

Comment (jonkeane) [reply r/inst/build_arrow_static.sh:70]: Yeah, partially: I suspect that there were builds where it was triggered because the names of the variables lined up just right and that's where you saw that, but at least on the latest build I'm seeing the same number of "Looking for" on both.

But also like neal mentioned up top, setting the env var with `-e` for the runtime only will make sure it gets in the place we need it without having to worry about passing it around. (I'm going to add a bit more to my comment up there since it's mostly related to that...)

File: .env (hunk 3 above)
Comment (jonkeane) [reply .env:70]: And for here, I would do ~the opposite of what I said yesterday, we want something like `-e ARROW_DEPENDENCY_SOURCE=...` in the docker call since that's the environment variable that we advertise and use throughout Arrow to determine that source.

Yesterday I mentioned calling it something like `R_STATIC_DEP...` in the `.env` file, which if we needed to pass it there would be good to mark that it only was getting passed to the R builds and wasn't a parameter that would help on a python build. But Neal's right that we should just put it in the docker command in the GHA workflow file and not worry about passing it around in docker-compose or the .env file. Sorry for my misdirection there yesterday!

File: r/inst/build_arrow_static.sh (hunk 4 above)
Comment (thisisnic) [reply r/inst/build_arrow_static.sh]: OK, deleted these changes now then

File: r/configure [hunk 6]
```diff
@@ -268,6 +269,11 @@ else
   PKG_CFLAGS=""
 fi
 
+
+if [ "$ARROW_DEPENDENCY_SOURCE" == "AUTO" ]; then
```
[the remark this answers is not in the document]
Comment (nealrichardson) [reply r/configure:273]: Also, before L160 (the call to nixlibs.R), add something like:

```
if [ "${ARROW_DEPENDENCY_SOURCE}" = "" ]; then
  # TODO: BUNDLED is still default for now, but we plan to change it to AUTO
  ARROW_DEPENDENCY_SOURCE=BUNDLED; export ARROW_DEPENDENCY_SOURCE
fi
if [ "${ARROW_DEPENDENCY_SOURCE}" = "AUTO" ]; then
  which pkg-config >/dev/null 2>&1
  if [ $? -eq 0 ]; then
    export ARROW_DEPENDENCY_SOURCE=BUNDLED
    # echo something?
  fi
fi

Finally, you also need to edit build_arrow_static.sh to take ARROW_DEPENDENCY_SOURCE from the environment (default arg there BUNDLED)

**Review by nealrichardson — CHANGES_REQUESTED** [review]

File: r/configure [hunk 7]
```diff
@@ -80,7 +80,8 @@ elif [ "$INCLUDE_DIR" ] && [ "$LIB_DIR" ]; then
 else
   # Use pkg-config if available and allowed
   pkg-config --version >/dev/null 2>&1
-  if [ $? -eq 0 ] && [ "$ARROW_USE_PKG_CONFIG" != "false" ]; then
+  if [ $? -eq 0 ] && [ "$ARROW_USE_PKG_CONFIG" != "false" ] && [ "$ARROW_DEPENDENCY_SOURCE" == "AUTO" ]; then
```
[the remark this answers is not in the document]
Comment (nealrichardson) [reply r/configure:83]: However, you may want to pull Kou's suggestion from https://github.com/apache/arrow/pull/10404#issuecomment-871034191 in here (though I'm not certain about the `--static` part in this block, may have to experiment locally)

File: r/configure [hunk 8]
```diff
@@ -173,6 +188,10 @@ else
         BUNDLED_LIBS=`echo "$BUNDLED_LIBS" | sed -e "s/\\.a lib/ -l/g" | sed -e "s/\\.a$//" | sed -e "s/^lib/-l/" | tr '\n' ' ' | sed -e "s/ $//"`
         PKG_DIRS="-L`pwd`/$LIB_DIR"
 
+        if [ "$ARROW_DEPENDENCY_SOURCE" == "AUTO" ]; then
+          PKG_LIBS = "pkgconfig ${LIB_DIR} $PKG_LIBS"
```
Comment (nealrichardson) [inline r/configure:192]: I think you still need https://github.com/apache/arrow/pull/10710#discussion_r678380437 here, and also adding a comment explaining what's up would help our future selves.

File: r/inst/build_arrow_static.sh (hunk 2 above)
Comment (thisisnic) [reply r/inst/build_arrow_static.sh]: So atfer that, I ran this:
`ARROW_R_DEV=TRUE DEPENDENCY_SOURCE=AUTO docker run -it apache/arrow-dev:r-rhub-ubuntu-gcc-release-latest bash`

but couldn't find that script in the `arrow` directory.  Has something gone really wrong here?

Comment (nealrichardson) [reply r/inst/build_arrow_static.sh]: I think you want `ARROW_R_DEV=TRUE docker run -it apache/arrow-dev:r-rhub-ubuntu-gcc-release-latest -e ARROW_DEPENDENCY_SOURCE=AUTO /bin/bash` (get the right variable name, and pass it into the runtime environment with -e)

Comment (nealrichardson) [reply r/inst/build_arrow_static.sh]: oh you also need `-v $(pwd):/arrow` or something like that, that's what mounts the /arrow directory based on your local dir

Comment (thisisnic) [reply r/inst/build_arrow_static.sh]: Thanks Neal!

Comment (thisisnic) [reply r/inst/build_arrow_static.sh]: Just to clarify then, the reason why ARROW_DEPENDENCY_SOURCE needs to be passed in via `-e` but `ARROW_R_DEV` doesn't is:

- 'r' (defined in docker-compose.yml) is a service
- specifying environment variables ahead of calling `docker run` allows you to set env vars for running that service
- with `ARROW_R_DEV`, it doesn't matter whether you specify it via `-e` or via prefixing the call to `docker run` as it gets passed through anyway via the `environment` field of the `r` service
- with `ARROW_DEPENDENCY_SOURCE`,  it must be set via `-e` as there's nothing defined in the service to explicitly pass it through

File: r/configure (hunk 7 above)
[the remark this answers is not in the document]
Comment (nealrichardson) [reply r/configure:83]: I think these comments are still valid.

[the remark this answers is not in the document]
Comment (thisisnic) [reply r/configure:83]: I don't disagree!  I'm still working on this - converted it to a draft for the moment whilst I go read up on things as I realised I was just changing code as per all the helpful suggestions, but not understanding why.

**Comment by thisisnic** [comment]
@github-actions crossbow submit test-r-depsource

**Comment by thisisnic** [comment]
@github-actions crossbow submit test-r-depsource

File: r/inst/build_arrow_static.sh (hunk 2 above)
Comment (thisisnic) [reply r/inst/build_arrow_static.sh]: Now tested and working as expected when pkg-config isn't installed

**Comment by thisisnic** [comment]
@github-actions crossbow submit test-r-depsource

**Comment by thisisnic** [comment]
@github-actions crossbow submit test-r-depsource

**Comment by thisisnic** [comment]
@github-actions crossbow submit test-r-depsource

**Comment by thisisnic** [comment]
@github-actions crossbow submit test-r-depsource

File: r/inst/build_arrow_static.sh (hunk 4 above)
Comment (nealrichardson) [reply r/inst/build_arrow_static.sh:57]: Let's leave a default value here so that the script doesn't depend on being called from nixlibs.R

```suggestion
    -DARROW_DEPENDENCY_SOURCE=${ARROW_DEPENDENCY_SOURCE:-BUNDLED} \
```

File: dev/tasks/tasks.yml [hunk 9]
```diff
@@ -941,6 +941,10 @@ tasks:
   test-r-devdocs:
     ci: github
     template: r/github.devdocs.yml
+    
+  test-r-depsource:
+    ci: github
+    template: r/github.linux.libsource.yml
```
Comment (nealrichardson) [inline dev/tasks/tasks.yml:947]: I'm not sure we need a custom .yml for this, I think you can get the desired result by using an existing docker image we test on and adding 

```
flags: '-e ARROW_DEPENDENCY_SOURCE=AUTO'
```

like `test-r-gcc-11` below

File: dev/tasks/tasks.yml [hunk 10]
```diff
@@ -941,6 +941,10 @@ tasks:
   test-r-devdocs:
     ci: github
     template: r/github.devdocs.yml
+    
+  test-r-depsource:
```
Comment (nealrichardson) [inline dev/tasks/tasks.yml:945]:
```suggestion
  test-r-depsource-auto:
```

**Comment by thisisnic** [comment]
@github-actions crossbow submit test-r-depsource-auto

File: dev/tasks/tasks.yml (hunk 9 above)
Comment (thisisnic) [reply dev/tasks/tasks.yml]: Will do!

**Comment by thisisnic** [comment]
@github-actions crossbow submit test-r-depsource-auto

**Comment by thisisnic** [comment]
@github-actions crossbow submit test-r-depsource-auto

**Comment by thisisnic** [comment]
@github-actions crossbow submit test-r-depsource-auto

**Comment by thisisnic** [comment]
@github-actions crossbow submit test-r-depsource-auto test-r-depsource-bundled

**Comment by thisisnic** [comment]
@github-actions crossbow submit test-r-depsource-auto test-r-depsource-bundled

**Comment by thisisnic** [comment]
Ready for re-review now.  I included a task with each of BUNDLED and AUTO dependency sources in the previous test run so I could verify it's working by doing a diff of the logs.  It looks all good now, and so I have removed the BUNDLED one.

File: dev/tasks/tasks.yml [hunk 11]
```diff
@@ -941,6 +941,13 @@ tasks:
   test-r-devdocs:
     ci: github
     template: r/github.devdocs.yml
+    
+  test-r-depsource-auto:
+    ci: github
+    template: docker-tests/github.linux.yml
+    params:
+      flags: '-e ARROW_DEPENDENCY_SOURCE=AUTO -e ARROW_SOURCE_HOME="/arrow" -e LIBARROW_BUILD=TRUE'
+      image: ubuntu-r
```
Comment (nealrichardson) [inline dev/tasks/tasks.yml:950]: This isn't the image you want to use. Look at the build logs: it's not using the build script we ship with the R package, and it builds a shared library. So it's not exercising any of your new code.

Pick one of the images enumerated below this and add a job like it that has the `flags` provided. That will exercise `r/inst/build_arrow_static.sh`.

**Review by nealrichardson — CHANGES_REQUESTED** [review]

File: dev/tasks/tasks.yml (hunk 11 above)
Comment (thisisnic) [reply dev/tasks/tasks.yml:950]: Sorry, I don't understand what you mean by "has the `flags` provided"?

Comment (nealrichardson) [reply dev/tasks/tasks.yml:950]:
```
  test-r-depsource-auto:
    ci: azure
    template: r/azure.linux.yml
    params:
      r_org: rocker
      r_image: r-base
      r_tag: latest
      flags: '-e ARROW_DEPENDENCY_SOURCE=AUTO'
```

Then it looks like you'll need to add `{{ flags|default("") }}` around L53 in dev/tasks/r/azure.linux.yml so that the job picks them up since that parameter isn't defined in that template yet. (I think there's some refactoring we can do here to clean up these templates, but that's for another time.)

**Comment by thisisnic** [comment]
@github-actions crossbow submit test-r-depsource-auto

**Comment by thisisnic** [comment]
test-r-depsource-auto

**Comment by thisisnic** [comment]
@github-actions crossbow submit test-r-depsource-auto

**Comment by nealrichardson** [comment]
@github-actions crossbow submit -g r

**Review by nealrichardson — APPROVED** [review]
Thanks for sticking with this. I'm still not sure we have everything right here, but we can test more when we try changing the default later.

**Comment by nealrichardson** [comment]
@github-actions crossbow submit -g r

**Comment by nealrichardson** [comment]
@jonkeane can you help make sure that this isn't breaking any builds? crossbow keeps returning failures, I haven't had the chance to examine why. This PR shouldn't be changing the default behavior.

**Comment by jonkeane** [comment]
@github-actions crossbow submit -g r

File: r/configure [hunk 12]
```diff
@@ -173,6 +186,11 @@ else
         BUNDLED_LIBS=`echo "$BUNDLED_LIBS" | sed -e "s/\\.a lib/ -l/g" | sed -e "s/\\.a$//" | sed -e "s/^lib/-l/" | tr '\n' ' ' | sed -e "s/ $//"`
         PKG_DIRS="-L`pwd`/$LIB_DIR"
 
+        # Use pkg-config to do static linking of libarrow's dependencies
+        if [ "$ARROW_DEPENDENCY_SOURCE" = AUTO ]; then
```
Comment (jonkeane) [inline r/configure:190]:
```suggestion
        if [ "$ARROW_DEPENDENCY_SOURCE" = AUTO | "$ARROW_DEPENDENCY_SOURCE" = SYSTEM ]; then
```

**Review by jonkeane — COMMENTED** [review]
Ok, I think I finally figured out what's going on with the failing builds: the `ARROW_DEPENDENCY_SOURCE` variable is being set early on in the docker creation (one example: https://github.com/apache/arrow/blob/master/ci/docker/ubuntu-21.04-cpp.dockerfile#L109) and now the R package passes that on to the building process, which is erroring out when building statically. 

It's possible that we need to do what I suggested above and do either auto or system when we do `--libs-only-other` and then it will work.

Thought we probably _also_ want to make sure that these builds are building what we think they are / building the default unset version. We could do this by adding `-e ARROW_DEPENDENCY_SOURCE=""` to the flags, though that is a lot of copy and paste.

File: r/configure (hunk 12 above)
Comment (thisisnic) [reply r/configure]: Nice find, makes a lot more sense now!

**Comment by thisisnic** [comment]
@github-actions crossbow submit -g r

File: dev/tasks/docker-tests/github.linux.yml [hunk 13]
```diff
@@ -41,7 +41,7 @@ jobs:
         shell: bash
         run: archery docker run -e SETUPTOOLS_SCM_PRETEND_VERSION="{{ arrow.no_rc_version }}" {{ flags|default("") }} {{ image }} {{ command|default("") }}
 
-    {% if '-r-' in image %}
+    {% if '-r' in image %}
       - name: Dump R install logs
         run: cat arrow/r/check/arrow.Rcheck/00install.out
         continue-on-error: true
```
Comment (jonkeane) [inline dev/tasks/docker-tests/github.linux.yml:47]:
```suggestion
        continue-on-error: true
        if: always()
```

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-ubuntu-21.04 test-r-gcc-11

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-ubuntu-21.04 test-r-gcc-11

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-ubuntu-21.04 test-r-gcc-11

**Comment by jonkeane** [comment]
Ok, I've got a few more pieces of what's going on here:

For 21.04 and gcc11 (which is based on 21.04). It looks like not _all_ of the system dependencies are being linked correctly in the R project. 

For example, thrift is installed in these images, and [is found when libarrow is being build](https://github.com/ursacomputing/crossbow/runs/3279031937?check_suite_focus=true#step:8:128), however, it's [not among the paths that are linked](https://github.com/ursacomputing/crossbow/runs/3279031937?check_suite_focus=true#step:8:1038), and [so when it is attempted to be loaded it errors](https://github.com/ursacomputing/crossbow/runs/3279031937?check_suite_focus=true#step:7:6874). This isn't happening with the `ARROW_DEPENDENCY_SOURCE=AUTO` test only because we hadn't pre-installed thrift on the system, it's not found during building and thus is build from source. But if someone did use `ARROW_DEPENDENCY_SOURCE=AUTO` and happen to have thrift installed, they would see this same issue.

What we need to do is, for the dependencies that are found, run something like https://github.com/apache/arrow/blob/7ca6906454486d87a885175a16eaeb8211596072/r/configure#L191 on them where instead of `arrow` being at the end, it's the dependency name (e.g. `thrift`) so that those are added to `PKG_LIBS`. There might be a way to get this information from Arrow's build process (I thought I remember a comment about this earlier in the thread from Neal, but I can't find it anymore).

Additionally, something [is off about the AWS SDK in the 21.04 build ](https://github.com/ursacomputing/crossbow/runs/3279031495?check_suite_focus=true#step:8:176)— it's not triggered by the gcc-11 build only because we've shut off S3 support (and therefore don't build the AWS SDK).  I'm not certain that this issue is a problem with this PR specifically, or something more general that should be solved elsewhere. I wouldn't be opposed to adding `ARROW_S3=OFF` to the 21.04 build as well for now and making a follow up jira to figure out what's going on there — we already have one for bumping the AWS SDK version to something that is compatible with GCC11, and that might fix this entirely.

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-ubuntu-21.04 test-r-gcc-11

**Comment by thisisnic** [comment]
@kou - we're trying to work out how to proceed with this ticket and had a question you might know the answer to - I've summarised everything below so you don't have to go trawling through all the history here!

The overall aim of this ticket is to enable the R package build to use pkg-config to see if the dependencies are already installed and link to those instead of the bundled ones when the R package is built/installed.  Currently, the R package only uses the bundled versions of the dependencies (i.e. ARROW_DEPENDENCY_SOURCE is set to BUNDLED) but we’re trying to make it work with that variable set to AUTO which will use either system versions if present or the bundled ones if not.

We're having a few issues around getting pkg-config working correctly in the R package - we realised that at the moment, in the code which builds the R package, pkg-config isn't actually being run against the various dependencies - though it is being used successfully to find libarrow.  A possible approach could be running pkg-config against every dependency when we build the R package in order to see if the package is already installed and get all the libraries directories and libraries link flags if it is, but that seems a bit inefficient.  

Given that cmake is already doing something similar and likely has  these flags when libarrow is being built, is there some way we can extract these flags from cmake when it builds libarrow to use when the R package is being built so we're not repeating ourselves?

**Comment by thisisnic** [comment]
Actually, wait, I think the answer may be in the JIRA ticket:
>  Alternatively, we could try to grep/sed it from the *.pc files.

File: r/configure [hunk 14]
```diff
@@ -173,6 +186,11 @@ else
         BUNDLED_LIBS=`echo "$BUNDLED_LIBS" | sed -e "s/\\.a lib/ -l/g" | sed -e "s/\\.a$//" | sed -e "s/^lib/-l/" | tr '\n' ' ' | sed -e "s/ $//"`
         PKG_DIRS="-L`pwd`/$LIB_DIR"
 
+        # Use pkg-config to do static linking of libarrow's dependencies
+        if [[ "$ARROW_DEPENDENCY_SOURCE" = "AUTO" || "$ARROW_DEPENDENCY_SOURCE" = "SYSTEM" ]]; then
```
Comment (kou) [inline r/configure:190]: We should use `[` not `[[` because `/bin/sh` may not be bash`: https://github.com/ursacomputing/crossbow/runs/3280854203?check_suite_focus=true#step:7:4328

```text
./configure: 190: [[: not found
./configure: 190: SYSTEM: not found
```

**Review by kou — COMMENTED** [review]
How about this?

```diff
diff --git a/cpp/cmake_modules/ThirdpartyToolchain.cmake b/cpp/cmake_modules/ThirdpartyToolchain.cmake
index 0631d277b..003d14428 100644
--- a/cpp/cmake_modules/ThirdpartyToolchain.cmake
+++ b/cpp/cmake_modules/ThirdpartyToolchain.cmake
@@ -837,7 +837,16 @@ set(Boost_ADDITIONAL_VERSIONS
 # so we first need to determine whether we're building it
 if(ARROW_WITH_THRIFT AND Thrift_SOURCE STREQUAL "AUTO")
   find_package(Thrift 0.11.0 MODULE COMPONENTS ${ARROW_THRIFT_REQUIRED_COMPONENTS})
-  if(NOT Thrift_FOUND AND NOT THRIFT_FOUND)
+  if(Thrift_FOUND)
+    pkg_check_modules(THRIFT_PC
+                      thrift
+                      NO_CMAKE_PATH
+                      NO_CMAKE_ENVIRONMENT_PATH
+                      QUIET)
+    if(THRIFT_PC_FOUND)
+      string(APPEND ARROW_PC_REQUIRES_PRIVATE " thrift")
+    endif()
+  else()
     set(Thrift_SOURCE "BUNDLED")
   endif()
 endif()
@@ -1306,7 +1315,7 @@ endmacro()
 if(ARROW_WITH_THRIFT)
   # We already may have looked for Thrift earlier, when considering whether
   # to build Boost, so don't look again if already found.
-  if(NOT Thrift_FOUND AND NOT THRIFT_FOUND)
+  if(NOT Thrift_FOUND)
     # Thrift c++ code generated by 0.13 requires 0.11 or greater
     resolve_dependency(Thrift
                        REQUIRED_VERSION
```

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-ubuntu-21.04 test-r-gcc-11

**Comment by thisisnic** [comment]
@github-actions crossbow submit test-r-ubuntu-21.04 test-r-gcc-11

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-ubuntu-21.04 test-r-gcc-11

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-ubuntu-21.04 test-r-gcc-11

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 15]
```diff
@@ -836,8 +836,18 @@ set(Boost_ADDITIONAL_VERSIONS
 # Thrift needs Boost if we're building the bundled version with version < 0.13,
 # so we first need to determine whether we're building it
 if(ARROW_WITH_THRIFT AND Thrift_SOURCE STREQUAL "AUTO")
+  find_package(PkgConfig QUIET)
```
Comment (jonkeane) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:839]: I think you added this @thisisnic ? Did it resolve the issue you thought it was going to? And do you remember what that error was?

File: dev/tasks/tasks.yml [hunk 16]
```diff
@@ -1013,6 +1013,15 @@ tasks:
...
     template: r/github.devdocs.yml
+    
+  test-r-depsource-auto:
+      ci: azure
+      template: r/azure.linux.yml
+      params:
+        r_org: rocker
+        r_image: r-base
+        r_tag: latest
+        flags: '-e ARROW_DEPENDENCY_SOURCE=AUTO'
```
Comment (jonkeane) [inline dev/tasks/tasks.yml:1024]: Should we try and add a task that also tests `ARROW_DEPENDENCY_SOURCE=SYSTEM`? To test when we can't fall back to building?  Maybe I'm paranoid, but I'm a bit worried if the rocker base happens to not have any of our dependencies installed, this isn't _actually_ exercising linking to them.

**Comment by jonkeane** [comment]
Ok, I just re-read through this to see if there were any extraneous things we introduced while debugging this. Thank you for all the persistence + help here @thisisnic I never would have found that missing `pkgconfig/` without your pointers about the .pc file.

The only thing left to resolve is how best should we unset `ARROW_DEPENDENCY_SOURCE` for builds based on ubuntu-r. We could punt on that and "just" add `-e ARROW_DEPENDENCY_SOURCE=""` to the flags for each of those tasks, but that's not the best solution.

File: dev/tasks/tasks.yml (hunk 16 above)
Comment (jonkeane) [reply dev/tasks/tasks.yml:1024]: We can also make a follow up for this, I don't think that `ARROW_DEPENDENCY_SOURCE=SYSTEM` will work with this image without extending it a bit to install the other dependencies we need.

**Comment by jonkeane** [comment]
@github-actions crossbow submit -g r

**Comment by jonkeane** [comment]
OH, oops forgot about the conda failures. Those seem to have a slightly different error (that RE2 is not found, though it looks like it is through the conda proces). We might need to have a switch to use the conda dependency resolution instead of pkgconfig there.

File: cpp/cmake_modules/ThirdpartyToolchain.cmake (hunk 15 above)
Comment (kou) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake:839]: This should be added just before `pkg_check_modules(...)`.

Comment (thisisnic) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake:839]: @jonkeane It is, and it fixes something, though as @kou says, it should be moved

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-depsource-auto test-r-depsource-system test-r-ubuntu-21.04

**Comment by jonkeane** [comment]
@github-actions crossbow submit conda-linux-gcc-py36-cpu-r40 conda-linux-gcc-py37-cpu-r41 conda-osx-clang-py36-r40 conda-osx-clang-py37-r41

**Comment by jonkeane** [comment]
@github-actions crossbow submit conda-linux-gcc-py36-cpu-r40 conda-linux-gcc-py37-cpu-r41 conda-osx-clang-py36-r40 conda-osx-clang-py37-r41 test-r-depsource-auto test-r-depsource-system

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 17]
```diff
@@ -837,7 +837,17 @@ set(Boost_ADDITIONAL_VERSIONS
 # so we first need to determine whether we're building it
 if(ARROW_WITH_THRIFT AND Thrift_SOURCE STREQUAL "AUTO")
```
Comment (jonkeane) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:838]:
```suggestion
if(ARROW_WITH_THRIFT AND (Thrift_SOURCE STREQUAL "AUTO" OR Thrift_SOURCE STREQUAL "SYSTEM"))
```

**Comment by jonkeane** [comment]
@github-actions crossbow submit conda-linux-gcc-py36-cpu-r40 conda-linux-gcc-py37-cpu-r41 conda-osx-clang-py36-r40 conda-osx-clang-py37-r41 test-r-depsource-auto test-r-depsource-system

**Comment by jonkeane** [comment]
@github-actions crossbow submit conda-linux-gcc-py36-cpu-r40 conda-linux-gcc-py37-cpu-r41 conda-osx-clang-py36-r40 conda-osx-clang-py37-r41 test-r-depsource-auto test-r-depsource-system

File: ci/docker/linux-apt-r.dockerfile [hunk 18]
```diff
@@ -100,7 +100,7 @@ ENV \
     ARROW_BUILD_STATIC=OFF \
     ARROW_BUILD_TESTS=OFF \
     ARROW_BUILD_UTILITIES=OFF \
-    ARROW_DEPENDENCY_SOURCE=SYSTEM \
+    ARROW_DEPENDENCY_SOURCE="" \
```
Comment (nealrichardson) [inline ci/docker/linux-apt-r.dockerfile:103]: Why this change? This build should not be affected by the scope of the changes in this PR.

File: dev/tasks/docker-tests/github.linux.yml [hunk 19]
```diff
@@ -41,10 +41,11 @@ jobs:
         shell: bash
         run: archery docker run -e SETUPTOOLS_SCM_PRETEND_VERSION="{{ arrow.no_rc_version }}" {{ flags|default("") }} {{ image }} {{ command|default("") }}
 
-    {% if '-r-' in image %}
+    {% if '-r' in image %}
```
Comment (nealrichardson) [inline dev/tasks/docker-tests/github.linux.yml:44]: Why this change? Is this too broad?

Comment (jonkeane) [reply dev/tasks/docker-tests/github.linux.yml:44]: Before this change, some builds (e.g. `test-r-ubuntu-21.04` which uses the docker-compose service `ubuntu-r-only-r` which extends `ubuntu-r`) didn't have the log dumping step showing up at all. https://github.com/ursacomputing/crossbow/runs/3272438137?check_suite_focus=true 

The strange thing is that it _does_ have `-r-` in the underlying image name (e.g. `${REPO}:${ARCH}-ubuntu-${UBUNTU}-r-${R}`), but somehow the jinja matching isn't quite getting it there (though `-r` works). It might be overly-broad though that step _should_ be harmless since it `continue-on-error`.

File: ci/docker/linux-apt-r.dockerfile (hunk 18 above)
Comment (jonkeane) [reply ci/docker/linux-apt-r.dockerfile:103]: This is that stacking of docker images + env vars I mentioned earlier. For any of our test jobs that use the `ubuntu-r` (and a few other linux distros) docker compose service, they will set the env var `ARROW_DEPENDENCY_SOURCE` to `SYSTEM` (e.g. https://github.com/apache/arrow/blob/master/ci/docker/ubuntu-21.04-cpp.dockerfile#L109). This is true even of our `ubuntu-r-only-r` task that extends (docker's term, not mine) the `ubuntu-r` service to *not* build libarrow first and only rely on the R package process to build libarrow. So all of those tests were no longer testing environments that most people would experience (with the `ARROW_DEPENDENCY_SOURCE` env var unset). 

I'll try moving this to the `environment` of the `ubuntu-r-only-r` service, since that's slightly more targeted and should work. But for a CI job that is intended to mimic what people are doing when they are installing (which was the original intention behind the `test-r-ubuntu-21.04` task, I think we want this to install with that env var unset.

**Comment by jonkeane** [comment]
@github-actions crossbow submit conda-linux-gcc-py36-cpu-r40 conda-linux-gcc-py37-cpu-r41 test-r-depsource-auto test-r-depsource-system test-r-ubuntu-21.04 test-r-gcc-11

**Comment by jonkeane** [comment]
@github-actions crossbow -g r

**Comment by jonkeane** [comment]
@github-actions crossbow submit -g r

**Comment by jonkeane** [comment]
AFAICT, conda-osx-clang-py36-r40, conda-osx-clang-py37-r41, homebrew-r-autobrew, test-r-rhub-debian-gcc-devel-lto-latest, and test-ubuntu-18.04-r-sanitizer are all failing on crossbow already independent of this change.

**Comment by nealrichardson** [comment]
Ok, I checked the `test-r-depsource-auto` build output. It does seem to be picking up system libz and libbz2, though that's it. We do have a few tests that are `skip_if_not_available("gzip")`, and those tests were not skipped, so that confirms that we've successfully used the system libz. 

So, +1 from me again on this. Will merge.
